### PR TITLE
cdc: add builtin function to emulate json encoding

### PR DESCRIPTION
**Pull request by HonoreDB**
There's been a [longstanding debugging request](https://github.com/cockroachdb/cockroach/issues/75730) to be able to see how a row would get encoded by a changefeed with certain options. The main obstacle was removing the need for the builtins package to depend on any package implementing a builtin, which got easier thanks to the introduction of the builtinsregistry. This PR builds on that work to make everything that references builtin functions load-order independent. The first commit removes the load order effect on function OIDs, and the second changes everything that iterates over all builtins to instead register a hook to be called on each builtin as it gets registered. Finally, we add a crdb_internal function that's registered and implemented outside of the builtins package.

Release note (sql change): Added the crdb_internal.to_json_as_changefeed_with_flags function to help
debug json changefeeds. 

Release justification: Improves stability by ensuring that version changes don't alter OIDs and helping with debugging.

### CODE REVIEW

**Comment by HonoreDB** [comment]
Sort of naturally composes with cdc expressions: `SELECT crdb_internal.to_json_as_changefeed_with_flags(projection.*, 'updated', 'diff') from [select address as place, 'dog' as hungry from users where id != 'ae147ae1-47ae-4800-8000-000000000022'] as projection limit 2;`

But we'd need very different function resolution logic to be able to use the cdc builtins there.

File: pkg/sql/sem/builtins/generator_builtins.go [hunk 1]
```diff
@@ -46,7 +46,7 @@ import (
 var _ eval.ValueGenerator = &seriesValueGenerator{}
 var _ eval.ValueGenerator = &arrayValueGenerator{}
 
-func initGeneratorBuiltins() {
+func init() {
```
Comment (miretskiy) [inline pkg/sql/sem/builtins/generator_builtins.go:49]: This is an okay change; but it probably should be done in a separate PR if you feel
strongly about this cleanup.

File: pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go [hunk 2]
```diff
@@ -20,16 +20,40 @@ import "github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 
 var registry = map[string]definition{}
 
+// Subscription is a hook to be called once on all registered builtins.
+type Subscription func(name string, props *tree.FunctionProperties, overloads []tree.Overload)
+
+var subscriptions = make([]Subscription, 0)
```
Comment (miretskiy) [inline pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go:26]: making subscription w/ 0 len is the same as `var subscriptions []Subscription`

File: pkg/ccl/utilccl/builtins.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+		Conversion: func(d tree.Datum) anyArg { return makeAnyArg(converter(d)) },
+	}
+}
+
+// RegisterCCLBuiltin converts an arbitrary go function to an equivalent SQL function
+// using reflection. Names should be namespaced (usually starting with `crdb_internal.ccl_`).
+// impl must be a function that returns a single type with a key in BuiltinReturnTypes,
+// plus optionally an error. This is a convenience method meant for generating SQL functions
+// that are tightly coupled to other behavior. It's generally preferable to create SQL builtins
+// via calling builtinsregistry.Register directly.
```
Comment (miretskiy) [inline pkg/ccl/utilccl/builtins.go:99]: I honestly don't know about this -- seems like too much magic.

As your comment says that it is "preferable to create and register via builtinsregistry"... why not make it required? What does this function add that builtinregistry doesn't provide in a very explicit way?

File: pkg/ccl/changefeedccl/cdcevent/event.go [hunk 4]
```diff
@@ -530,3 +532,29 @@ func TestingGetFamilyIDFromKey(
 	_, familyID, err := decoder.(*eventDecoder).rfCache.tableDescForKey(context.Background(), key, ts)
 	return familyID, err
 }
+
+// MakeRowFromTuple converts a SQL datum produced by, for example, SELECT ROW(foo.*),
+// into the same kind of cdcevent.Row you'd get as a result of an insert, but without
+// the primary key.
+func MakeRowFromTuple(t *tree.DTuple) Row {
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/cdcevent/event.go:539]: Very nice; this functions wants to be tested :)

File: pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go [hunk 5]
```diff
@@ -20,16 +20,40 @@ import "github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
...
+	for _, s := range subscriptions {
+		s(name, props, overloads)
+	}
+}
+
+// AddSubscription defines a hook to be called once on all registered builtins.
+// Subscriptions should be added at init() time, but are load-order independent:
+// if you add a subscription after a function is registered, it will immediately
+// be called on that function, while functions that are registered afterwards will
+// also trigger the hook.
```
Comment (miretskiy) [inline pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go:51]: having a bit of a hard time seeing why we need this.
Can you explain what is the purpose of this?
Does it make sense to pull it into separate commit?

File: pkg/ccl/changefeedccl/encoder_json.go [hunk 6]
```diff
@@ -195,3 +198,40 @@ func (e *jsonEncoder) EncodeResolvedTimestamp(
...
+	e, err := makeJSONEncoder(opts, changefeedbase.Targets{})
+	if err != nil {
+		return nil, err
+	}
+	return e.EncodeValue(context.TODO(), placeholderCtx, r, cdcevent.Row{})
+
+}
+
+func init() {
+	utilccl.RegisterArgType(cdcevent.Row{}, types.AnyTuple, func(d tree.Datum) interface{} {
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/encoder_json.go:230]: I just don't see how this is more readable than registering builtin where you
explicitly say that the function accepts a DTuple, and that you convert that tuple into a row, and then execute conversion.

File: pkg/sql/sem/tree/overload.go [hunk 7]
```diff
@@ -134,7 +134,7 @@ type Overload struct {
 	SQLFn SQLFnOverload
 
 	// OnTypeCheck is incremented every time this overload is type checked.
-	OnTypeCheck func()
+	OnTypeCheck *func()
```
Comment (miretskiy) [inline pkg/sql/sem/tree/overload.go:137]: I think this needs some explanation.. It's very infrequent that I see a function pointer...

**Review by miretskiy — CHANGES_REQUESTED** [review]

**Review by HonoreDB — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84509)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @shermanCRL)

___
*[`pkg/ccl/utilccl/builtins.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72q2hZcDy60yjp_zyy-r1-99:-N7GYikS24AII4a3ea9j:bj1vbao) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/ccl/utilccl/builtins.go#L99)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I honestly don't know about this -- seems like too much magic.

As your comment says that it is "preferable to create and register via builtinsregistry"... why not make it required? What does this function add that builtinregistry doesn't provide in a very explicit way?
</blockquote></details>

...yeah, I don't know if I can really defend this. My goal was to make it possible to add builtin functions without having to learn all the unusual idioms of SQL function definition code. Right now to define a changefeed-related function you need as an engineer to understand both changefeed code and SQL builtins code and you need to have a single file that depends on both `changefeedccl` and a bunch of `tree` objects. But once this file gets a few type mappings into it, you'll instead be able to write code that's normal for your package.

That said, I don't think "people should be able to write builtins without knowing what a tree.Overload is" is necessarily a good idea.
___
*[`pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01e-r1-51:-N7GWcEtEBdkIe_jPF95:btm0f9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go#L51)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

having a bit of a hard time seeing why we need this.
Can you explain what is the purpose of this?
Does it make sense to pull it into separate commit?
</blockquote></details>

The reason is that I don't know how to control load order without adding a dependency. There is probably a way? If I could guarantee that everything that needed to know about all builtins would load after everything that registered a builtin, I could just add a builtinsregistry.Freeze() function to make sure it stayed that way by panicking if you tried to register a builtin after calling a method like Iterate or AllBuiltinNames. But if we're going to register builtins in ccl packages, we can't have all_builtins take a dependency on every file that creates a builtin. The other other way to do it would be to have ccl builtins get registered no matter what and just throw an error in open source builds, but I feel like it's worth doing a little work to avoid that.
___
*[`pkg/sql/sem/tree/overload.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N73USCj0-wD2HYn_0yL-r1-137:-N7GXxXGC8DNjxoBLAfK:b-7e16sh) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/tree/overload.go#L137)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think this needs some explanation.. It's very infrequent that I see a function pointer...
</blockquote></details>

I bet there was a better way to do this--I'll add a comment, but the idea was to be able to define this function in an AddSubscription hook when the overload is passed by value.

**Review by HonoreDB — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84509)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @rafiss, and @shermanCRL)

___
*[`pkg/ccl/changefeedccl/encoder_json.go` line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01Z-r1-230:-N8AlWme97yMQXV9OBew:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/ccl/changefeedccl/encoder_json.go#L230)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I just don't see how this is more readable than registering builtin where you
explicitly say that the function accepts a DTuple, and that you convert that tuple into a row, and then execute conversion.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/cdcevent/event.go` line 539 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01X-r1-539:-N8AkdEtD6WfOwVnEb2J:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/ccl/changefeedccl/cdcevent/event.go#L539)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Very nice; this functions wants to be tested :)
</blockquote></details>

Added.
___
*[`pkg/ccl/utilccl/builtins.go` line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72q2hZcDy60yjp_zyy-r1-99:-N8AkfzSAkrfYjt2Zs3e:b-t98xe2) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/ccl/utilccl/builtins.go#L99)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

...yeah, I don't know if I can really defend this. My goal was to make it possible to add builtin functions without having to learn all the unusual idioms of SQL function definition code. Right now to define a changefeed-related function you need as an engineer to understand both changefeed code and SQL builtins code and you need to have a single file that depends on both `changefeedccl` and a bunch of `tree` objects. But once this file gets a few type mappings into it, you'll instead be able to write code that's normal for your package.

That said, I don't think "people should be able to write builtins without knowing what a tree.Overload is" is necessarily a good idea.
</blockquote></details>

I have sighed wistfully and deleted all the magic.
___
*[`pkg/sql/crdb_internal.go` line 1578 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N8AkwdA1kCY4arDLIAm:-N8AkwdA1kCY4arDLIAn:b-tbhavb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f786aadcad04f3bfb7d7dce2bf138531d8d91f7c/pkg/sql/crdb_internal.go#L1578)):*
> ```Go
> 			}
> 
> 			if !hasModify && !hasView {
> ```

Reminder to self to make sure this diff isn't actually in the PR.
___
*[`pkg/sql/sem/builtins/all_builtins.go` line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01c-r1-56:-N8Al3lb6nPPeqndp2s4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/builtins/all_builtins.go#L56)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

It would be nice to pull such refactoring into a separate commit (or even PR).
I'm pretty sure schema folks would have something to say about this.

I would prefer to have a better understanding (through nice commit message) why this change is needed (as I see it, it's because you want to add more things to those builtins, without adding the builtin implementation in this directory (and keeping it it under CCL).  But again, the mechanism for doing this is up for debate (or at least comments/approval from schema people).
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N73d91R_2JrSvnCj-5K-r1-49:-N8Al6xO21X2MFcpA7c3:bkwyng0) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/builtins/generator_builtins.go#L49)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This is an okay change; but it probably should be done in a separate PR if you feel
strongly about this cleanup.
</blockquote></details>

I've pulled it into a separate commit. It sort of needs to be in this PR in order to prove that registration is build-order-independent.
___
*[`pkg/sql/sem/tree/overload.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N73USCj0-wD2HYn_0yL-r1-137:-N8AlPwpEmqr8t9i9wTr:bkchjt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/tree/overload.go#L137)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I bet there was a better way to do this--I'll add a comment, but the idea was to be able to define this function in an AddSubscription hook when the overload is passed by value.
</blockquote></details>

(comment added)

**Review by HonoreDB — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84509)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @rafiss, and @shermanCRL)

___
*[`pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go` line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-N72kbtU_dJDuWEmu01e-r1-26:-N8AmrR02cCIPS8oF0Rv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31388508b2c67a74423124a7485d56e018b435d6/pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go#L26)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

making subscription w/ 0 len is the same as `var subscriptions []Subscription`
</blockquote></details>

Done.

File: pkg/ccl/changefeedccl/cdcevent/event.go [hunk 8]
```diff
@@ -530,3 +532,29 @@ func TestingGetFamilyIDFromKey(
...
+		} else {
+			name = names[i]
+		}
+		r.AddValueColumn(name, d.ResolvedType())
+		if err := r.SetValueDatumAt(i, d); err != nil {
+			if build.IsRelease() {
+				log.Warningf(context.Background(), "failed to set row value from tuple due to error %v", err)
+				_ = r.SetValueDatumAt(i, tree.DNull)
+			} else {
+				panic(err)
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/cdcevent/event.go:555]: do we have to panic? Seems like returning error is fine?

File: pkg/ccl/changefeedccl/encoder_json.go [hunk 9]
```diff
@@ -195,3 +201,62 @@ func (e *jsonEncoder) EncodeResolvedTimestamp(
...
+			return tree.NewDBytes(tree.DBytes(o)), nil
+		},
+		Info: "Strings can be of the form 'resolved' or 'resolved=1s'.",
+		// Probably actually stable, but since this is tightly coupled to changefeed logic by design,
+		// best to be defensive.
+		Volatility: volatility.Volatile,
+	}
+
+	utilccl.RegisterCCLBuiltin("crdb_internal.to_json_as_changefeed_with_flags",
+		`Encodes a tuple the way a changefeed would output it if it were inserted as a row or emitted by a changefeed expression, and returns the raw bytes.
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/encoder_json.go:259]: `crdb_internal.cdc_encode_as_json`?
Or even `cdc_to_json` -- kinda similiar to `row_to_json`?

**Review by miretskiy — APPROVED** [review]
cdc stuff -- fine; the rest, somebody need to stamp it.

**Comment by HonoreDB** [comment]
This is ready for a look on the SQL end, first two commits are builtins registration refactors.

**Review by rafiss — APPROVED** [review]
this lgtm, but @chengxiong-ruan may want to look at the first 2 commits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-:-NAk-gUE1Tjky6ubSJJf:bg0wkra)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @rafiss)

___
*[`pkg/sql/sem/builtins/all_builtins.go` line 32 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-NAjLaEk1hh4oYT3wMKy:-NAjLaEk1hh4oYT3wMKz:b-50pdm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bdf8236a508a808b660f045df72755d4a473afb/pkg/sql/sem/builtins/all_builtins.go#L32)):*
> ```Go
> }
> 
> func (o *orderedStrings) add(s string) {
> ```

this could use comments saying the differences between `add` and `insert`
___
*[`pkg/sql/sem/builtins/all_builtins.go` line 47 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/84509#-NAjLlmI492gyqxNUqxY:-NAjLlmI492gyqxNUqxZ:bjlgg30) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bdf8236a508a808b660f045df72755d4a473afb/pkg/sql/sem/builtins/all_builtins.go#L47)):*
> ```Go
> }
> 
> func (o *orderedStrings) insert(s string) {
> ```

if you call `insert` on a non-sorted list, should it become sorted first?

File: pkg/sql/sem/builtins/fixed_oids.go [hunk 10]
```diff
@@ -0,0 +1,2041 @@
...
+	`voidin(input: anyelement) -> void`:                                                                                                                                             1997,
+	`voidout(void: void) -> bytes`:                                                                                                                                                  1996,
+	`voidrecv(input: anyelement) -> void`:                                                                                                                                           1995,
+	`voidsend(void: void) -> bytes`:                                                                                                                                                 1994,
+	`width_bucket(operand: decimal, b1: decimal, b2: decimal, count: int) -> int`:                                                                                                   811,
+	`width_bucket(operand: int, b1: int, b2: int, count: int) -> int`:                                                                                                               812,
+	`width_bucket(operand: anyelement, thresholds: anyelement[]) -> int`:
```
Comment (chengxiong-ruan) [inline pkg/sql/sem/builtins/fixed_oids.go:2038]: nit: just return `existing max id + 1`, what do you think?
I guess the effort here is trying get any unused id in between, and that'd be helpful if we removed any builtin functions. I'd prefer `existing max id + 1` because: (1) removing builtin function could be super rare. (2) reuse an id seems not desirable since we're trying to make OIDs deterministic here.

File: pkg/sql/sem/builtins/all_builtins.go [hunk 11]
```diff
@@ -14,63 +14,88 @@ import (
...
+	o.strings[i] = s
+}
+
+var allBuiltinNames orderedStrings
+
+// AllBuiltinNames returns a slice containing all the built-in function
 // names, sorted in alphabetical order. This can be used for a
 // deterministic walk through the Builtins map.
-var AllBuiltinNames []string
+func AllBuiltinNames() []string {
```
[the remark this answers is not in the document]
Comment (chengxiong-ruan) [reply pkg/sql/sem/builtins/all_builtins.go:30]: +1, I'm not sure why this function is preferred over a sorted string slice. Trying make the string slice private maybe? If so, I think we may just sort it once in the `init` function right?

**Review by chengxiong-ruan — APPROVED** [review]
LGTM, but curious with 2 questions.

File: pkg/sql/sem/builtins/fixed_oids.go (hunk 10 above)
Comment (HonoreDB) [reply pkg/sql/sem/builtins/fixed_oids.go:2037]: Good point, done.

File: pkg/sql/sem/builtins/all_builtins.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (HonoreDB) [reply pkg/sql/sem/builtins/all_builtins.go:30]: This actually just sorts the slice once--further calls to .sort() will short-circuit. It's happening here rather than in an init() so that we don't depend on load order to fully populate the slices before they're sorted.

[the remark this answers is not in the document]
Comment (chengxiong-ruan) [reply pkg/sql/sem/builtins/all_builtins.go:30]: I see, thanks, for some reason I missed the sortedStrings implementation above.

**Comment by HonoreDB** [comment]
bors r=[miretskiy,rafiss,chengxiong-ruan]
